api/v1beta1/azuremachine_webhook.go

```diff
@@ -152,6 +154,24 @@ func (m *AzureMachine) ValidateUpdate(oldRaw runtime.Object) error {
 		}
 	}
 
+	if !reflect.DeepEqual(m.Spec.NetworkInterfaces, old.Spec.NetworkInterfaces) {
+		// The defaulting webhook may have migrated values from the old SubnetName field to the new NetworkInterfaces format.
+		old.Spec.SetNetworkInterfacesDefaults()
+
+		// The reconciler will populate the SubnetName on the first interface if the user left it blank.
+		if old.Spec.NetworkInterfaces[0].SubnetName == "" && m.Spec.NetworkInterfaces[0].SubnetName != "" {
+			old.Spec.NetworkInterfaces[0].SubnetName = m.Spec.NetworkInterfaces[0].SubnetName
+		}
```

Comment:
The controller itself patches the SubnetName field if it is empty, so we have to allow that specific mutation. This is for backwards compatibility. The defaulting webhook cannot set the SubnetName reliably, because we cannot guarantee that the AzureCluster has been persisted  when the defaulting webhook runs for the AzureMachine.

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2411/files#diff-b6931ac61f37a13f66fc7836e039bf714a0553d5db4f313cc8ef31ca5f5647b9R710-R713